From 8b0b3825f183663d2ce852c09bfe1f20c4943630 Mon Sep 17 00:00:00 2001
From: Michael Trensch <mtrensch@hilscher.com>
Date: Thu, 17 Dec 2020 10:20:18 +0100
Subject: [PATCH] This patch adds all enabled input ports into
 FWDI_<zone>_allow.

It requires an additional dependency on FORWARD_IN for all calls to "backend.build_zone_ports_rules", as otherwise
all rules will fail as FWDI does not yet exist

It additionally modifies mangle/PREROUTING to mark all incoming packets on closed ports.
0x10000 = reject
0x20000 = drop

Dropping and rejecting is done in DOCKER-USER chain which is run as early as possible
---
 src/firewall/core/fw_zone.py   | 10 ++++++--
 src/firewall/core/ipXtables.py | 38 ++++++++++++++++++++++------
 src/firewall/core/nftables.py  | 45 +++++++++++++++++++++++++++-------
 3 files changed, 75 insertions(+), 18 deletions(-)

diff --git a/src/firewall/core/fw_zone.py b/src/firewall/core/fw_zone.py
index a62a58b8..520f82d3 100644
--- a/src/firewall/core/fw_zone.py
+++ b/src/firewall/core/fw_zone.py
@@ -1608,7 +1608,8 @@ class FirewallZone(object):
 
                     # create rules
                     for (port,proto) in svc.ports:
-                        if enable and type(rule.action) == Rich_Mark:
+                        if enable:
+                            zone_transaction.add_chain("filter", "FORWARD_IN")
                             zone_transaction.add_chain("mangle", "PREROUTING")
                         rules = backend.build_zone_ports_rules(
                                     enable, zone, proto, port, destination, rule)
@@ -1637,7 +1638,7 @@ class FirewallZone(object):
 
                 if enable:
                     zone_transaction.add_chain("filter", "INPUT")
-                if enable and type(rule.action) == Rich_Mark:
+                    zone_transaction.add_chain("filter", "FORWARD_IN")
                     zone_transaction.add_chain("mangle", "PREROUTING")
 
                 rules = backend.build_zone_ports_rules(
@@ -1772,6 +1773,8 @@ class FirewallZone(object):
                     modules.append(nat_module)
                 zone_transaction.add_modules(modules)
             zone_transaction.add_chain("filter", "INPUT")
+            zone_transaction.add_chain("filter", "FORWARD_IN")
+            zone_transaction.add_chain("mangle", "PREROUTING")
 
         # build a list of (backend, destination). The destination may be ipv4,
         # ipv6 or None
@@ -1825,6 +1828,8 @@ class FirewallZone(object):
     def _port(self, enable, zone, port, protocol, zone_transaction):
         if enable:
             zone_transaction.add_chain("filter", "INPUT")
+            zone_transaction.add_chain("filter", "FORWARD_IN")
+            zone_transaction.add_chain("mangle", "PREROUTING")
 
         for backend in self._fw.enabled_backends():
             if not backend.zones_supported:
@@ -1877,6 +1882,7 @@ class FirewallZone(object):
 
         if enable:
             zone_transaction.add_chain("nat", "PREROUTING")
+            zone_transaction.add_chain("mangle", "PREROUTING")
 
         if toaddr and enable:
             zone_transaction.add_post(enable_ip_forwarding, ipv)
diff --git a/src/firewall/core/ipXtables.py b/src/firewall/core/ipXtables.py
index 77e70c66..d0d1cfd9 100644
--- a/src/firewall/core/ipXtables.py
+++ b/src/firewall/core/ipXtables.py
@@ -863,6 +863,14 @@ class ip4tables(object):
 
         target = self._fw.zone._zones[zone].target
 
+        if table == "mangle" and chain == "PREROUTING":
+            if target in [ "REJECT", "%%REJECT%%" ]:
+                rules.append([ "-A", "%s_pre" % _zone, "-t", table,
+                               "-j", "MARK", "--set-mark", "0x10000"])
+            if target == "DROP":
+                rules.append([ "-A", "%s_pre" % _zone, "-t", table,
+                               "-j", "MARK", "--set-mark", "0x20000"])
+
         if self._fw.get_log_denied() != "off":
             if table == "filter" and \
                chain in [ "INPUT", "FORWARD_IN", "FORWARD_OUT", "OUTPUT" ]:
@@ -1044,6 +1052,8 @@ class ip4tables(object):
         table = "filter"
         target = DEFAULT_ZONE_TARGET.format(chain=SHORTCUTS["INPUT"],
                                             zone=zone)
+        fwd_target = DEFAULT_ZONE_TARGET.format(chain=SHORTCUTS["FORWARD_IN"],
+                                            zone=zone)
 
         rule_fragment = [ "-p", proto ]
         if port:
@@ -1057,13 +1067,22 @@ class ip4tables(object):
             rule_fragment += [ "-m", "conntrack", "--ctstate", "NEW,UNTRACKED" ]
 
         rules = []
-        if rich_rule:
-            rules.append(self._rich_rule_log(rich_rule, enable, table, target, rule_fragment))
-            rules.append(self._rich_rule_audit(rich_rule, enable, table, target, rule_fragment))
-            rules.append(self._rich_rule_action(zone, rich_rule, enable, table, target, rule_fragment))
-        else:
-            rules.append([add_del, "%s_allow" % (target), "-t", table] +
-                         rule_fragment + [ "-j", "ACCEPT" ])
+        for tmp_target in [target, fwd_target]:
+            if rich_rule:
+                rules.append(self._rich_rule_log(rich_rule, enable, table, tmp_target, rule_fragment))
+                rules.append(self._rich_rule_audit(rich_rule, enable, table, tmp_target, rule_fragment))
+                rules.append(self._rich_rule_action(zone, rich_rule, enable, table, tmp_target, rule_fragment))
+            else:
+                rules.append([add_del, "%s_allow" % (tmp_target), "-t", table] +
+                             rule_fragment + [ "-j", "ACCEPT" ])
+
+        if port:
+            pre_target = DEFAULT_ZONE_TARGET.format(chain=SHORTCUTS["PREROUTING"],
+                                                zone=zone)
+            rules.append([add_del, "%s_allow" % (pre_target), "-t", "mangle"] +
+                         [ "-p", proto ] +
+                         [ "--dport", "%s" % portStr(port) ] +
+                         [ "-j", "MARK", "--set-mark", "0/0x30000" ])
 
         return rules
 
@@ -1202,6 +1221,11 @@ class ip4tables(object):
                      ["-p", protocol, "--dport", portStr(port),
                       "-j", "DNAT", "--to-destination", to])
 
+        # This is an allowed port, thus remove marks for packet rejection
+        rules.append(["-t", "mangle", add_del, "%s_%s" % (target, chain_suffix),
+                      "-p", protocol, "--dport", portStr(port),
+                      "-j", "MARK", "--set-mark", "0/0x30000"])
+
         return rules
 
     def build_zone_icmp_block_rules(self, enable, zone, ict, rich_rule=None):
diff --git a/src/firewall/core/nftables.py b/src/firewall/core/nftables.py
index 0b40604d..321fd1b1 100644
--- a/src/firewall/core/nftables.py
+++ b/src/firewall/core/nftables.py
@@ -700,6 +700,16 @@ class nftables(object):
 
         target = self._fw.zone._zones[zone].target
 
+        if table == "mangle" and chain == "PREROUTING":
+            if target in ["%%REJECT%%", "REJECT"]:
+                    rules.append(["add", "rule", family, "%s" % TABLE_NAME,
+                                  "%s_%s" % ("mangle", _zone),
+                                  "mark", "set", "0x10000"])
+            if target == "DROP":
+                    rules.append(["add", "rule", family, "%s" % TABLE_NAME,
+                                  "%s_%s" % ("mangle", _zone),
+                                  "mark", "set", "0x20000"])
+
         if self._fw.get_log_denied() != "off":
             if table == "filter" and \
                chain in ["INPUT", "FORWARD_IN", "FORWARD_OUT", "OUTPUT"]:
@@ -944,6 +954,7 @@ class nftables(object):
         add_del = { True: "add", False: "delete" }[enable]
         table = "filter"
         target = DEFAULT_ZONE_TARGET.format(chain=SHORTCUTS["INPUT"], zone=zone)
+        fwd_target = DEFAULT_ZONE_TARGET.format(chain=SHORTCUTS["FORWARD_IN"], zone=zone)
 
         rule_fragment = []
         if rich_rule:
@@ -962,14 +973,21 @@ class nftables(object):
             rule_fragment += ["ct", "state", "new,untracked"]
 
         rules = []
-        if rich_rule:
-            rules.append(self._rich_rule_log(rich_rule, enable, table, target, rule_fragment))
-            rules.append(self._rich_rule_audit(rich_rule, enable, table, target, rule_fragment))
-            rules.append(self._rich_rule_action(zone, rich_rule, enable, table, target, rule_fragment))
-        else:
-            rules.append([add_del, "rule", "inet", "%s" % TABLE_NAME,
-                          "%s_%s_allow" % (table, target)] +
-                          rule_fragment + ["accept"])
+        for tmp_target in [target, fwd_target]:
+            if rich_rule:
+                rules.append(self._rich_rule_log(rich_rule, enable, table, tmp_target, rule_fragment))
+                rules.append(self._rich_rule_audit(rich_rule, enable, table, tmp_target, rule_fragment))
+                rules.append(self._rich_rule_action(zone, rich_rule, enable, table, tmp_target, rule_fragment))
+            else:
+                rules.append([add_del, "rule", "inet", "%s" % TABLE_NAME,
+                              "%s_%s_allow" % (table, tmp_target)] +
+                              rule_fragment + ["accept"])
+
+        # This is an allowed port, thus remove marks for packet rejection
+        rules.append([add_del, "rule", "inet", "%s" % TABLE_NAME,
+                      "%s_%s_allow" % ("mangle", pre_target)] +
+                      [proto, "dport", "%s" % portStr(port, "-")] +
+                      ["mark", "set", "0"])
 
         return rules
 
@@ -1141,12 +1159,21 @@ class nftables(object):
         else:
             chain_suffix = "allow"
 
-        return [[add_del, "rule", family, "%s" % TABLE_NAME,
+        rules = [[add_del, "rule", family, "%s" % TABLE_NAME,
                 "nat_%s_%s" % (target, chain_suffix)]
                 + rule_fragment +
                 [protocol, "dport", port]
                 + dnat_fragment]
 
+        # This is an allowed port, thus remove marks for packet rejection
+        rules.append([add_del, "rule", "inet", "%s" % TABLE_NAME,
+                      "%s_%s_allow" % ("mangle", pre_target)] +
+                      [protocol, "dport", port] +
+                      ["mark", "set", "0"])
+
+        return rules
+
+
     def build_zone_forward_port_rules(self, enable, zone, port,
                                       protocol, toport, toaddr, rich_rule=None):
         rules = []
-- 
2.17.1

